context.go

```diff
@@ -157,21 +176,37 @@ func merge[K comparable, V any](a, b map[K][]V) (merged map[K][]V) {
 	return
 }
 
+// encodeOneAddressType encodes the given addressData values and returns the corresponding WAF object and its refs.
+// If the addressData is empty, it returns nil for the WAF object and an empty ref pool.
+// At this point, the only error we can get is an error in case the top level object is a nil map, but this
+// behaviour is expected since either persistent or ephemeral addresses are allowed to be null one at a time.
+// In this case, EncodeAddresses will return nil contrary to Encode which will return an nil wafObject,
+// which is what we need to send to ddwaf_run to signal that the address data is empty.
+func (context *Context) encodeOneAddressType(addressData map[string]any, runTimings *Timings) (*bindings.WafObject, *encoder, error) {
+	var data *bindings.WafObject = nil
+	encoder := newLimitedEncoder(runTimings.Remaining())
+	if addressData != nil {
+		data, _ = encoder.EncodeAddresses(addressData)
+		runTimings.EncoderNs.Add(encoder.watch.Spent().Nanoseconds())
+	}
+
+	if runTimings.Expired() {
+		return nil, nil, errors.ErrTimeout
+	}
+
+	return data, &encoder, nil
+}
+
 // run executes the ddwaf_run call with the provided data on this context. The caller is responsible for locking the
 // context appropriately around this call.
 func (context *Context) run(persistentData, ephemeralData *bindings.WafObject, timeout time.Duration) (Result, error) {
 	result := new(bindings.WafResult)
 	defer wafLib.WafResultFree(result)
 
-	ret := wafLib.WafRun(context.cContext, persistentData, ephemeralData, result, uint64(timeout/time.Microsecond))
+	ret := wafLib.WafRun(context.cContext, persistentData, ephemeralData, result, uint64(timeout.Microseconds()))
 
-	context.totalRuntimeNs.Add(result.TotalRuntime)
-	res, err := unwrapWafResult(ret, result)
-	if err == errors.ErrTimeout {
-		context.timeoutCount.Inc()
-	}
-
-	return res, err
+	context.timings.RuntimeNs.Add(int64(result.TotalRuntime))
```

Comment:
`uint64` to `int64` conversion feels gross (the other way around is "fine" if you can guarantee it's never negative).